### DISCUSSION TITLE
Performance improvements for connected component computations

### DISCUSSION THREAD

**Discussion by michaeldorner** [networkx/networkx#4997] [General]
The computation of the connected component in networkx.algorithms.components.connected relies on a BFS. The function is called `_plain_bfs(...)`: 

```
def _plain_bfs(G, source):
    """A fast BFS node generator"""
    G_adj = G.adj
    seen = set()
    nextlevel = {source}
    while nextlevel:
        thislevel = nextlevel
        nextlevel = set()
        for v in thislevel:
            if v not in seen:
                seen.add(v)
                nextlevel.update(G_adj[v])
    return seen
```

I have benchmarked this code (please find [my script as gist](https://gist.github.com/michaeldorner/7298678eef1e2ab35e94a4de1f74dfc8)) and found that the following code is faster for all generated random graphs:

```
def bfs(G, source):
    seen = {source}
    next = {source}
    while next:
        v = next.pop()
        for n in G.neighbors(v):
            if n not in seen:
                seen.add(n)
                next.add(n)
    return seen
```

I personally find it more readable, too. Replacing this function would improve the performance of all related connected component functions. Does it make sense to replace it? Have I missed something?

**Comment by michaeldorner** [comment]
I also noticed that `descendants` and `ancestors ` use `shortest_path_length`. Since we are not using any weights, what about using the BFS implementation above as well for this candidates?

**Comment by dschult** [comment]
This looks like an improvement to me.  I haven't tested the timing of the two, but if adding each node is faster than appending and then updating the set once, it should be faster. That might depend on version of Python. I'll try to get a test of that. 

And I think you rare right that the `ancestors` and `descendants` functions should use the `plain_bfs` implementation too.  

There may be places in the heart of `shortest_paths` that could use this same trick -- anywhere we don't need the levels to be identified could simply `add` each node instead of appending to a list for that level and then updating the set.

**Reply by michaeldorner** [reply]
Shall I draft a pull request starting with replacing the `plain_bfs`?

**Reply by dschult** [reply]
Yes please! :}

**Reply by dschult** [reply]
Hmmm.... Maybe I spoke too quickly.  
The routine you propose isn't a BFS algorithm... right?
I think the order is not even known due to `set` order difficulties.
We need to figure out if that matters or not for these routines.

**Reply by michaeldorner** [reply]
I’m not sure but I would say both implementations are not „proper“ BFS because both functions return a set which has no order anyway. Shall we call the „BFS“ differently? What about `horizon` of a node to indicate it’s not a ordered result of a BFS?

**Reply by dschult** [reply]
The original code doesn't order within each level, but it is a BFS because it works one level at a time out from the source node.  That is, nodes are encountered in the order of increasing distance to the source with ties in that order being resolved arbitrarily (by the set object order).

**Reply by michaeldorner** [reply]
You are right. But as the order does not matter for the component and the in-level order gets lost anyway, we could replace the existing code by a faster one. But let me know if you think differently.

**Comment by dschult** [comment]
@boothby has done some timing experiments with various versions of node search to see which are better under what conditions.  They report (verbally to me) that 5 different versions each are faster for graphs with different traits.  What type of graphs did you test your node-search code on?  Is it worth extending the tests to a wider variety?  @boothy do you have suggestions for how to vary the graphs we time?

**Reply by michaeldorner** [reply]
I used 10 different random graphs generated by

    G = nx.gnm_random_graph(random.randint(1,1000), random.randint(1,1000))

Please find my mini benchmark [here](https://gist.github.com/michaeldorner/7298678eef1e2ab35e94a4de1f74dfc8).

**Comment by boothby** [comment]
Thanks, Michael.  With some more detailed analysis, I've found that your approach is really good for graphs with many components, but the current strategy is better for graphs with few components.  I've gone through and optimized both approaches, and come up with a blended approach which starts with the breadth-first search for the first component, and then switches over to the 'ad hoc' (calling it that because it's not properly breadth- or depth-first) search for the remaining components.

On the test set I've used (including both dense and sparse graphs), the performance of the blended approach is almost always within 10% of the "optimal" algorithm.  I've done a breakdown of the statistics into graphs with few/many components, because the apparent optimality in the whole set depends on the ratio of graphs with few to graphs with many components.

https://gist.github.com/boothby/97e9e9ca228b93271da04736269a311f

And, the blended approach is kinda ugly.  I'll break it out here; there's more implementations, with explanations, in the gist above.

```python
def connected_components(G):
    if not G:
        return

    ichain = chain.from_iterable
    adj = G._adj.__getitem__

    todo = set(G)
    todo_pop = todo.pop
    todo_done = todo.difference_update

    component = set()
    nextlevel = {todo_pop()}
    component_contains = component.__contains__
    component_update = component.update
    while nextlevel:
        lastlevel = nextlevel
        nextlevel = set(ichain(map(adj, filterfalse(component_contains, lastlevel))))
        component_update(lastlevel)

    todo_done(component)
    yield component

    next = set()
    next_update = next.update 
    next_pop = next.pop 
    while todo:
        v = todo_pop()
        component = {v} 
        next_update(adj(v))
        if next: #this greatly speeds up singleton components
            component_add = component.add 
            while next: 
                u = next_pop() 
                if u not in component:
                    component_add(u) 
                    next_update(adj(u))
            todo_done(component) 
        yield component
```

In the end, I think this is worthwhile.  On graphs with few components, this implementation takes 2/3 of the time of the current implementation.  On graphs with many components, it's several times faster.

**Reply by michaeldorner** [reply]
Excellent work, @boothby! Thanks for your efforts. 

I rerun your benchmark and reformatted it as pandas DataFrame to get a table overview:

| graph  | implementation        | absolute            | relative            |
|--------|-----------------------|---------------------|---------------------|
| sparse | connected_components2 | 0.13874903199803157 | 0.27520126997201727 |
| sparse | connected_components6 | 0.14896725600078753 | 0.29546856972843566 |
| sparse | connected_components1 | 0.1903194300002724  | 0.3774883909616643  |
| sparse | connected_components5 | 0.23339302599947587 | 0.462922560487334   |
| sparse | connected_components4 | 0.27156291900087126 | 0.5386304978863427  |
| sparse | connected_components0 | 0.5002462880017902  | 0.9922117060883947  |
| sparse | connected_components3 | 0.5041729350018613  | 1.0                 |
| dense  | connected_components6 | 0.89453858599677    | 0.5766480818252726  |
| dense  | connected_components4 | 0.918588141996679   | 0.5921511920914933  |
| dense  | connected_components3 | 1.134743914001774   | 0.7314920916942848  |
| dense  | connected_components0 | 1.1603719429997454  | 0.7480127359616031  |
| dense  | connected_components2 | 1.1923417349971714  | 0.7686214827728169  |
| dense  | connected_components1 | 1.37476113499838    | 0.88621484179064    |
| dense  | connected_components5 | 1.5512729759982449  | 1.0                 |

When comparing the relative speed-ups (in comparison to the slowest approach -- independent who that is) for sparse and dense graphs (equally weighted), we see that 

| implementation        | relative |
|-----------------------|----------|
| connected_components6 | 0.436058 |
| connected_components2 | 0.521911 |
| connected_components4 | 0.565391 |
| connected_components1 | 0.631852 |
| connected_components5 | 0.731461 |
| connected_components3 | 0.865746 |
| connected_components0 | 0.870112 |

**Reply by michaeldorner** [reply]
Although @boothby's code is excellently written and optimized, and the fastest approach, I am unsure if `connected_components6` is the best choice in terms of readability and maintainability. 

IMHO, the proposed ad-hoc solution (without the really smart optimizations) provides the best balance between readability, maintainability, and performance improvement in all dimensions to the current state. `connected_components6` further improves the performance, but at the expenses of readability and maintainability (again IMHO).

What do you think, @boothby and @dschult?

**Reply by boothby** [reply]
I've added another column, which examines the total time spent by each algorithm in solving all problems.  Relative time was interesting for showing where the successes are happening, but obscures the difference between runtimes of dense and sparse graphs.  Since the algorithms all take time like O(N+E), it's important to look at what happens when E = O(n^2) (which describes "most graphs" but that's a bit of a philosophical point).  And there, even the most-optimized ad-hoc search doesn't do so hot.  It doesn't even compete with the current implementation.

I've updated my gist, the table below is abbreviated from there.

## Overall time: 1569 trials
| implementation | total time |
| -------------- | ----------:|
| connected_components0 | 35.076 |
| connected_components1 | 77.654 |
| connected_components2 | 57.434 |
| connected_components3 | 34.758 |
| connected_components4 | 39.002 |
| connected_components5 | 126.818 |
| connected_components6 | 30.190 |
| connected_components7 | 30.551 |
| connected_components8 | 30.796 |

https://gist.github.com/boothby/97e9e9ca228b93271da04736269a311f

So, in my opinion, we could take one of the inlined, functional implementations of BFS.  It could use some comments to explain what's going on to facilitate readability & maintenance.  Or, we could leave it be.  The runtime is a small improvement, about 6/7 of the current implementaion.

**Reply by michaeldorner** [reply]
Thanks for your work. 

1. O(n^2) is just for fully connected graphs. I wouldn’t say this is the standard case. 
2. In my opinion, the total time is not too meaningful: it has a bias towards dense graphs since they just take longer. 

But, of course, leave it as it is is fine with me.

**Reply by boothby** [reply]
I disagree with 1 -- there's a lot of non-complete graphs which have a number of edges proportional to n^2 -- we should be using big theta here, not big O.

Regardless, after some further refinement of the data, I've found that my initial assessment was inaccurate -- it isn't necessarily the number of components that's important, but rather, the average degree does appear to play a role.  And what I've found is that the ad-hoc search wins for average degree 2, but loses to the current implementation for average degree above 5 or so.  It might be fun to fit the runtime of these algorithms against a model `a*num_nodes + b*num_edges + c*num_components`, but I won't be putting that effort in myself.

I'm firm on my opinion that the ad-hoc search is not a straightforward performance improvement.  I think it's fair to characterize it as making fast problems go faster, and slow problems go slower.  That agrees with your observation about the bias in the total time, but I do think that's an acceptable bias.  The marginal improvement of the fast problems is quite overwhelmed by the slowdown incurred by the slow problems.

There's a bit of a dilemma that library-writers face.  We can never really know what the tool will be used for.  We don't know the sizes, densities, or structures of the graphs that our users use, unless they give us that feedback (and we aren't about to start collecting statistics).  I think a pragmatic approach is to keep an eye on the worst case, and for this algorithm, that means paying attention to the behavior on dense graphs.
